File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Builder/Builder.ts
interface User {
  name: string;
  age: number;
  money: number;
  isEmployee: boolean;
  hasJob: boolean;
}
class UserBuilderTs {
  private name: string = "";
  private age: number = 0;
  private money: number = 0;
  private hasJob: boolean = false;
  private isEmployee: boolean = false;
  setName(name: string): UserBuilderTs {
    this.name = name;
    return this;
  }

  setAge(age: number): UserBuilderTs {
    this.age = age;
    return this;
  }

  setMoney(money: number): UserBuilderTs {
    this.money = money;
    return this;
  }

  makeJob(): UserBuilderTs {
    this.hasJob = true;
    return this;
  }

  makeEmployee(): UserBuilderTs {
    this.hasJob = true;
    this.isEmployee = true;
    return this;
  }
  build(): User {
    return {
      name: this.name,
      age: this.age,
      money: this.money,
      isEmployee: this.isEmployee,
      hasJob: this.hasJob,
    };
  }
}

class Person {
  public name: string;
  public age: number;
  public money: number;
  public isEmployee: boolean;
  public hasJob: boolean;
  constructor(builder: User) {
    this.name = builder.name;
    this.age = builder.age;
    this.money = builder.money;
    this.hasJob = builder.hasJob;
    this.isEmployee = builder.isEmployee;
  }
}
const person: Person = new Person(
  new UserBuilderTs().setName("mahdi").setAge(19).setMoney(200).build()
);
console.log(person);

File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Builder/Builder.py
class UserBuilder :
    def __init__(self):
     self._name = ''
     self._age = 0
     self._hasJob = False
     self._money = 0
     self._isEmployee = False
    def setName(self,name):
        self._name = name
        return  self
    def setAge(self,age):
        self._age = age
        return self
    def setMoney(self,money):
        self._money = money
        return self
    def makeEmployee(self):
        self._hasJob = True
        self._isEmployee = True
        return self
    def build(self):
        return {
            "name":self._name,
            "age":self._age,
            "money" : self._money,
            "isEmployee" : self._isEmployee,
            "hasJob" : self._hasJob
        }
class Person:
    def __init__(self,builder):
        self.name = builder.name
        self.age = builder.age
        self.money = builder.money
        self.hasJob = builder.hasJob
        self.isEmployee = builder.isEmployee

person = Person(UserBuilder().setName('mahdi').setAge(25).setMoney(200).build())
print(person)